### DISCUSSION TITLE
rabbitmqctl cannot Force exchange queues to quorum queue

### DISCUSSION THREAD

**Discussion by LangJV** [rabbitmq/rabbitmq-server#11988] [Other] [answered]
### Describe the bug

https://github.com/rabbitmq/rabbitmq-server/pull/10663 Seems to have added support for a "queue-type" or "queue_type" policy to facilitate exchange federation queues using quorum queues instead of classic mirrored queues. This is good, as it helps us fully and finally migrate off of classic mirrored queues in preparation for 4.0

However, this seems to not be a workable parameter to pass via rabbitmqctl when attempting to configure this. It rejects it as "not a valid parameter".

Example CLI command we are attempting:

```
[jlang1_alt@rabapdvw201 ~]$ sudo /usr/sbin/rabbitmqctl set_policy -p is-inventory --priority 20 --apply-to queues fedqueues .*-fed$ '{"federation-upstream-set":"all","queue-type":"quorum","ha-mode":"exactly","ha-params":2,"ha-sync-mode":"automatic","ha-sync-batch-size":5000,"queue-master-locator":"min-masters","ha-promote-on-failure":"always","ha-promote-on-shutdown":"when-synced","max-length":1500000,"max-length-bytes":1050000000}'
Setting policy "fedqueues" for pattern ".*-fed$" to "{"federation-upstream-set":"all","queue-type":"quorum","ha-mode":"exactly","ha-params":2,"ha-sync-mode":"automatic","ha-sync-batch-size":5000,"queue-master-locator":"min-masters","ha-promote-on-failure":"always","ha-promote-on-shutdown":"when-synced","max-length":1500000,"max-length-bytes":1050000000}" with priority "20" for vhost "is-inventory" ...
Error:
Validation failed

[{<<"queue-type">>,<<"quorum">>}] are not recognised policy settings

[jlang1_alt@rabapdvw201 ~]$ 
```

It does this when passing both "queue_type" and "queue-type", and "x-queue-type"

Is there like a "word bank" somewhere that needs to be updated to list this as a valid policy setting to set? I browsed the code but couldn't find what i was expecting to see in that vein in the rabbitmqctl code...

We are running on RHEL 8.10, a 5 node cluster using the RPM installation, currently RabbitMQ 3.13.2 and 3.13.6 have been tried with the same results. Everything appears to work otherwise, and all other "parameters" passed in the above command are working as expected. If we remove the "queue-type" value, the command will succeed.

### Reproduction steps

```
[jlang1_alt@rabapdvw201 ~]$ sudo /usr/sbin/rabbitmqctl set_policy -p is-inventory --priority 20 --apply-to queues fedqueues .*-fed$ '{"federation-upstream-set":"all","queue-type":"quorum","ha-mode":"exactly","ha-params":2,"ha-sync-mode":"automatic","ha-sync-batch-size":5000,"queue-master-locator":"min-masters","ha-promote-on-failure":"always","ha-promote-on-shutdown":"when-synced","max-length":1500000,"max-length-bytes":1050000000}'
Setting policy "fedqueues" for pattern ".*-fed$" to "{"federation-upstream-set":"all","queue-type":"quorum","ha-mode":"exactly","ha-params":2,"ha-sync-mode":"automatic","ha-sync-batch-size":5000,"queue-master-locator":"min-masters","ha-promote-on-failure":"always","ha-promote-on-shutdown":"when-synced","max-length":1500000,"max-length-bytes":1050000000}" with priority "20" for vhost "is-inventory" ...
Error:
Validation failed

[{<<"queue-type">>,<<"quorum">>}] are not recognised policy settings

[jlang1_alt@rabapdvw201 ~]$ 
```

### Expected behavior

My expected behavior is being able to set this new policy setting via rabbitmqctl so we can "convert" all of our exchange federation queues to be "quorum" type instead of "classic mirrored" type.

**Comment by mkuratczyk** [comment, marked as answer]
You are trying to apply a queue policy, but you can't set a queue type through a policy like that.

https://github.com/rabbitmq/rabbitmq-server/pull/10663 added a `queue-type` option to the exchange federation policy:
https://www.rabbitmq.com/docs/federation-reference#applying-to-federated-exchanges-only

Keep in mind that the you need to be careful during the transition: make sure you understand what will happen if you just add this parameter to your federation policy (what will happen to existing queues, messages in them, will there be a gap in what's federated when no queue exists briefly, etc). The details depend on other parameters and the exact process.

Side note: you could have migrated this to quorum queues long time ago - the queue doesn't have to be declared by the federation.

**Reply by mkuratczyk** [reply]
One more thing - when you set the queue type to `quorum`, remove all those mirroring related parameters (`ha-*`)

**Reply by LangJV** [reply]
@mkuratczyk - THANK YOU appreciate the information. In hindsight - setting it "just" on the federation policy does make sense. We had been attempting to set it on the "queue" policy because it seemed like a queue setting (queue-type) - whoops!

Regarding the migration to quorum queues - we have migrated all of our "normal" queues to quorum a long time ago. I wasn't aware there was a way to get our federated exchanged queues (which are greyed out in the console, and automatically created) to be quorum earlier than this. Either way - this should "convert" them and ensure they are always quorum going forward

We will remove the ha- and other classic policy settings too!

**Reply by michaelklishin** [reply]
Policy documentation clearly states that policies exist for settings that change at runtime. You cannot change the type of a queue at runtime. It is constant once set.

**Reply by LangJV** [reply]
@michaelklishin - are you saying we cannot set queue-type: quorum to control our federated exchange queue type then? Or are you simply saying a queue-type parameter set on a "queue" policy wouldn't make logical sense because a queue type cannot be changed after creation, and policies are for "changeable things"

**Reply by michaelklishin** [reply]
@LangJV consider [reading documentation](https://www.rabbitmq.com/docs/parameters#policies). The very second sentence in the Policies section explains that queue type cannot be set via a policy.

You cannot change queue type at runtime. You cannot therefore set queue type via a policy. You can limit to what queue or stream types a policy applies by specifying the type in the `apply-to` key.

**Reply by LangJV** [reply]
@michaelklishin - thank you for bearing with me. I think I'm getting this now. I need to apply this as a parameter to my federation definition. I think this shouldn't be too difficult...
I'm not sure why i got so wrapped around on this, but appreciate the patience (and shoving me the same link repeatedly until I actually read it!! )

**Reply by damianVarela** [reply]
I'm having the same issue. 

Versions: RabbitMQ 4.1.3 on Erlang 27.3.4.2
Community Docker image on a Debian host.

These are the commands I'm running (in a .sh script for IaC), and a dedicated user for federation links following POLP which authenticates with mTLS. Everything works, except the queue type. 

```
#!/bin/bash

# setup_rabbitmq.sh
# Usage: ./setup_rabbitmq.sh <admin_username> <admin_password>

if [ $# -ne 2 ]; then
    echo "Usage: $0 <admin_username> <admin_password>"
    exit 1
fi

ADMIN_USER=$1
ADMIN_PASS=$2

echo "Setting up rabbitmq cluster..."

# Create virtual host
rabbitmqctl add_vhost prod --description "Production environment" --default-queue-type quorum --tags prod

# Verify admin credentials and set permissions
rabbitmqctl authenticate_user "$ADMIN_USER" "$ADMIN_PASS"
rabbitmqctl set_permissions -p prod "$ADMIN_USER" ".*" ".*" ".*"

# Create users (passwordless - using TLS/X.509 authentication) (this will be in all clusters)
rabbitmqctl add_user snapshot-upstream-link ""
rabbitmqctl clear_password snapshot-upstream-link

# Set regular permissions for users (order: CONFIG WRITE READ)
rabbitmqctl set_permissions -p prod snapshot-upstream-link "^federation: realtime\.snapshot ->.*" "^federation: realtime\.snapshot ->.*" "^(realtime\.snapshot|federation: realtime\.snapshot ->.*)$"

# Set topic permissions
rabbitmqctl set_topic_permissions -p prod snapshot-producer realtime.snapshot "^realtime\..*\.snapshot$" ""

# Create exchange using rabbitmqadmin
rabbitmqadmin -u "$ADMIN_USER" -p "$ADMIN_PASS" --vhost "prod" declare exchange name="realtime.snapshot" type="direct" durable="true"

# Create federation upstreams with all URIs and proper TLS parameters (sample of just one of them)
rabbitmqctl set_parameter -p prod federation-upstream realtime-upstream '{"uri":["amqps://rabbitmq.server1.awesome-company.com:5671/prod?cacertfile=/certs/intermediate/ca-chain.cert.pem&certfile=/certs/clients/snapshot-upstream-link/snapshot-upstream-link.cert.pem&keyfile=/certs/clients/snapshot-upstream-link/snapshot-upstream-link.key.pem&verify=verify_peer&auth_mechanism=external&server_name_indication=rabbitmq.server1.awesome-company.com"]}'

# Create federation upstream set with correct syntax (sample)
rabbitmqctl set_parameter -p prod federation-upstream-set realtime-upstream-set '[{"upstream":"realtime-upstream"}]'

# Create federation policy and this is where i get the error
rabbitmqctl set_policy -p prod realtime-snapshot-federation-policy "^realtime\.snapshot$" '{"federation-upstream-set":"realtime-upstream-set", "queue-type":"quorum"}' --apply-to exchanges
```

if a remove the queue type key, it works, but then, all queues are classic ones. is it possible to make them quorum? I even expect it to have them as quorum as that is what I set the default queue type for the vhost.

Thanks in advance for your time.

**Reply by mkuratczyk** [reply]
The `queue-type` parameter should be a part of the `rabbitmqctl set_parameter -p prod federation-upstream ...` command.

**Reply by damianVarela** [reply]
Thanks! That was my mistake! I really appreciate your assistance.
